lib/ask/runtime/channel_broker_state.ex

```diff
@@ -80,201 +58,234 @@ defmodule Ask.Runtime.ChannelBrokerState do
+
+    Repo.update_all(query, [])
+
+    Repo.delete_all(
+      from q in Queue,
+        where:
+          q.channel_id == ^state.channel_id and q.respondent_id == ^respondent_id and
+            not is_nil(q.contacts) and q.contacts <= 0
+    )
+
+    state
   end
 
+  # Deactivates a contact and removes them from the queue.
   def deactivate_contact(state, respondent_id) do
-    respondent_contacts = contacts_for(state, respondent_id)
-
-    if respondent_contacts > 1 do
-      update_active_contact(state, respondent_id, fn active_contact ->
-        active_contact
-        |> Map.put(:contacts, respondent_contacts - 1)
-        |> Map.put(:last_contact, SystemTime.time().now)
-      end)
-    else
-      state
-      |> Map.put(:active_contacts, Map.delete(state.active_contacts, respondent_id))
-    end
+    Repo.delete_all(
+      from q in Queue,
+        where: q.channel_id == ^state.channel_id and q.respondent_id == ^respondent_id
+    )
+
+    state
   end
 
-  # Returns the current number of active contacts for the respondent.
-  defp contacts_for(state, respondent_id) do
-    case Map.get(state.active_contacts, respondent_id) do
-      %{contacts: contacts} -> contacts
-      _ -> 0
-    end
+  # Deactivates a contact and puts them back into the queue.
+  def reenqueue_contact(state, respondent_id, priority \\ :normal) do
+    query =
+      from q in Queue,
+        where: q.channel_id == ^state.channel_id and q.respondent_id == ^respondent_id
+
+    Repo.update_all(query,
+      set: [
+        queued_at: SystemTime.time().now,
+        priority: priority,
+        contacts: nil,
+        last_contact: nil,
+        channel_state: nil
+      ]
+    )
+
+    state
   end
 
+  # Returns true if we can active a new contact: there are pending contacts in
+  # queue that we can activate, and the channel is under capacity.
   def can_unqueue(state) do
-    cond do
-      PQueue.empty?(state.contacts_queue) -> false
-      activable_contacts?(state.contacts_queue) -> under_capacity?(state)
-      true -> false
+    if Queue.activable_contacts?(state.channel_id) do
+      under_capacity?(state)
+    else
+      false
     end
   end
 
-  defp activable_contacts?(queue) do
-    PQueue.any?(queue, fn [_, item] ->
-      case item do
-        {_, _, not_before, _} -> Date.compare(not_before, SystemTime.time().now) != :gt
-        _ -> true
-      end
-    end)
-  end
-
+  # OPTIMIZE: Cache the number of active contacts into the state to heavily
+  #           reduce the number of COUNT(*) queries that tend to be slow.
+  #
+  #           on load: count from the database & cache the value
+  #           on activate: increment the number (by size)
+  #           on increment/decrement/deactivate/reenqueue: update the count
   defp under_capacity?(state) do
-    count_active_contacts(state) < state.capacity
+    Queue.count_active_contacts(state.channel_id) < state.capacity
   end
```

Comment:
This method may have a negative impact on performance: we're counting how many contacts are active for the channel right from the database, and we count every time we try to activate a contact.

We could keep a counter in `state` for the number of active contacts, and manipulate it as we activate / increment / decrement / deactivate / reenqueue, instead of counting all the time. We'd only count from the database on `init/1`.

Reply:
Yes, there are a couple indexes:

- PRIMARY KEY (channel_id, respondent_id)
- INDEX (last_contact)

I tried to avoid aggregate methods everywhere, preferring the quick `Repo.exists?`, but the capacity check is the only request where I needed an actual value.

I didn't implement the optimization because keeping a cache is easy, but keeping it correctly updated is another matter.